osu.Game/Screens/Ranking/SoloResultsScreen.cs

```diff
@@ -1,90 +1,130 @@
+                // this simplifies handling later.
+                if (clonedScore.Equals(Score) || clonedScore.MatchesOnlineID(Score))
                 {
-                    var score = scores.Scores[i];
-                    int position = i + 1;
+                    Score.Position = clonedScore.Position;
+                    sortedScores.Add(Score);
+                }
+                else
+                    sortedScores.Add(clonedScore);
+            }
+
+            // if we haven't encountered a match for the presented score, we still need to attach it.
+            // note that the above block ensuring that the `Score` reference makes it in here makes this valid to write in this way.
+            if (!sortedScores.Contains(Score))
+                sortedScores.Add(Score);
 
-                    if (score.MatchesOnlineID(Score))
+            sortedScores = sortedScores.OrderByTotalScore().ToList();
+
+            int delta = 0;
+            bool isPartialLeaderboard = leaderboardManager.CurrentCriteria?.Scope != BeatmapLeaderboardScope.Local && result.TopScores.Count >= 50;
+
+            for (int i = 0; i < sortedScores.Count; i++)
+            {
+                var sortedScore = sortedScores[i];
+
+                if (!isPartialLeaderboard)
+                    sortedScore.Position = i + 1;
+                else
+                {
+                    if (ReferenceEquals(sortedScore, Score) && sortedScore.Position == null)
                     {
-                        // we don't want to add the same score twice, but also setting any properties of `Score` this late will have no visible effect,
-                        // so we have to fish out the actual drawable panel and set the position to it directly.
-                        var panel = ScorePanelList.GetPanelForScore(Score);
-                        Score.Position = panel.ScorePosition.Value = position;
+                        int? previousScorePosition = i > 0 ? sortedScores[i - 1].Position : 0;
+                        int? nextScorePosition = i < result.TopScores.Count - 1 ? sortedScores[i + 1].Position : null;
+
+                        if (previousScorePosition != null && nextScorePosition != null && previousScorePosition + 1 == nextScorePosition)
+                        {
+                            sortedScore.Position = previousScorePosition + 1;
+                            delta += 1;
+                        }
+                        else
+                            sortedScore.Position = null;
                     }
                     else
-                    {
-                        var converted = score.ToScoreInfo(rulesets, Beatmap.Value.BeatmapInfo);
-                        converted.Position = position;
-                        toDisplay.Add(converted);
-                    }
+                        sortedScore.Position += delta;
```

Comment:
This is the specific block of code lifted from https://github.com/ppy/osu/pull/32947. Note that while the general idea is the same, there are some differing details.